File: src/simulation/space/generator/RandomMapGenerator.java
package simulation.space.generator;

import extra.SpaceProbabilityFuncs;
import kotlin.collections.ArraysKt;
import kotlin.random.Random;
import simulation.space.*;
import simulation.space.resource.Genome;
import simulation.space.resource.Resource;
import simulation.space.resource.dependency.ResourceDependency;
import simulation.space.resource.dependency.ResourceDependencyDep;

import java.util.*;
import java.util.stream.Collectors;

import static shmp.random.RandomCollectionsKt.*;
import static shmp.random.RandomSpaceKt.randomTile;

/**
 * Basic map generator.
 */
public class RandomMapGenerator {
    public static WorldMap generateMap(int x, int y, int platesAmount, Random random) {
        List<List<Tile>> tiles = createTiles(x, y);
        WorldMap map = new WorldMap(tiles);
        setTileNeighbours(map);
        List<TectonicPlate> tectonicPlates = randomPlates(
                platesAmount,
                map,
                random
        );
        tectonicPlates.forEach(map::addPlate);
        fill(map);
        return map;
    }

    public static void fillResources(
            WorldMap map,
            ResourcePool resourcePool,
            MapGeneratorSupplement supplement,
            Random random
    ) {
        for (Resource resource : resourcePool.getResourcesWithPredicate(r ->
                r.getSpreadProbability() != 0 || r.getGenome().getType().equals(Genome.Type.Mineral))) {
            scatter(
                    map,
                    resourcePool,
                    resource,
                    random.nextInt(
                            supplement.getStartResourceAmountRange().getStart(),
                            supplement.getStartResourceAmountRange().getEndInclusive()
                    ),
                    random
            );
        }
    }

    private static void setTileNeighbours(WorldMap map) {
        int x = map.getX();
        int y = map.getY();
        for (int i = 0; i < x; i++) {
            for (int j = 0; j < y; j++) {
                map.get(i, j).setNeighbours(
                        Arrays.stream(new Tile[]{
                                map.get(i, j + 1),
                                map.get(i, j - 1),
                                map.get(i + 1, j),
                                map.get(i - 1, j)
                        })
                                .filter(Objects::nonNull)
                                .collect(Collectors.toList())
                );
            }
        }
    }

    private static List<List<Tile>> createTiles(int x, int y) {
        List<List<Tile>> map = new ArrayList<>();
        for (int i = 0; i < x; i++) {
            map.add(new ArrayList<>());
            for (int j = 0; j < y; j++) {
                map.get(i).add(new Tile(i, j));
            }
        }
        return map;
    }

    private static List<TectonicPlate> randomPlates(int platesAmount, WorldMap map, Random random) {
        List<TectonicPlate> tectonicPlates = new ArrayList<>();
        Set<Tile> usedTiles = new HashSet<>();
        for (int i = 0; i < platesAmount; i++) {
            TectonicPlate.Direction direction = randomElement(
                    ArraysKt.toList(TectonicPlate.Direction.values()),
                    random
            );
            TectonicPlate.Type type = randomElement(
                    ArraysKt.toList(TectonicPlate.Type.values()),
                    random
            );
            TectonicPlate tectonicPlate = new TectonicPlate(direction, type);
            Tile tile = randomTile(map, random);
            tectonicPlate.add(tile);
            tectonicPlates.add(tectonicPlate);
            usedTiles.add(tile);
        }
        boolean sw = true;
        while (sw) {
            sw = false;
            for (Territory territory : tectonicPlates) {
                List<Tile> brink = territory.getBrinkWithCondition(t -> !usedTiles.contains(t));
                if (brink.isEmpty()) {
                    continue;
                }
                Tile tile = randomElement(brink, random);
                territory.add(tile);
                usedTiles.add(tile);
                sw = true;
            }
        }
        return tectonicPlates;
    }

    private static void fill(WorldMap map) {
        boolean sw = true, ssw = true;
        for (TectonicPlate plate : map.getTectonicPlates()) {
            if (sw) {
                plate.setType(TectonicPlate.Type.Terrain);
                sw = false;
            } else if (ssw) {
                plate.setType(TectonicPlate.Type.Oceanic);
                ssw = false;
            }
            plate.initialize();
        }
        map.platesUpdate();
    }

    private static void scatter(WorldMap map, ResourcePool resourcePool, Resource resource, int n, Random random) {
        List<Tile> goodTiles = map.getTilesWithPredicate(t -> resource.getGenome().isOptimal(t));
        for (int i = 0; i < n; i++) {
            Tile tile;
            if (goodTiles.isEmpty()) {
                tile = randomTile(map, random);
                while (!resource.getGenome().isAcceptable(tile)) {
                    tile = randomTile(map, random);
                }
            } else {
                tile = randomElement(goodTiles, random);
            }
            tile.addDelayedResource(resource.copy());
            addDependencies(resource, tile, resourcePool);
        }
    }

    private static void addDependencies(Resource resource, Tile tile, ResourcePool resourcePool) {
        for (ResourceDependency dependency : resource.getGenome().getDependencies()) {
            if (!dependency.isPositive() || !dependency.isResourceNeeded()) {
                continue;
            }
            if (dependency instanceof ResourceDependencyDep && ((ResourceDependencyDep) dependency).getResourceNames()
                    .stream().anyMatch(s -> s.equals("Vapour"))) {
                return;
            }
            if (dependency instanceof ResourceDependencyDep) {
                for (String name : ((ResourceDependencyDep) dependency).getResourceNames()) {
                    Resource dep = resourcePool.getResource(name);
                    if (dep.getGenome().isAcceptable(tile)) {
                        tile.addDelayedResource(dep);
                    }
                    addDependencies(dep, tile, resourcePool);
                }
                for (String name : ((ResourceDependencyDep) dependency).getMaterialNames()) {
                    for (Resource dep : resourcePool.getResourcesWithPredicate(r ->
                            r.getSpreadProbability() > 0
                                    && !r.getSimpleName().equals(resource.getSimpleName())
                                    && r.getGenome().getPrimaryMaterial() != null
                                    && r.getGenome().getPrimaryMaterial().getName().equals(name))) {
                        if (dep.getGenome().isAcceptable(tile)) {
                            tile.addDelayedResource(dep.copy());
                            addDependencies(dep, tile, resourcePool);
                        }
                    }
                }
            }
        }
    }
}

File: src/simulation/space/generator/RandomMapGenerator.kt
package simulation.space.generator

import shmp.random.randomElement
import shmp.random.randomTile
import simulation.space.ResourcePool
import simulation.space.TectonicPlate
import simulation.space.Tile
import simulation.space.WorldMap
import simulation.space.resource.Genome
import simulation.space.resource.Resource
import simulation.space.resource.dependency.ResourceNeedDependency
import java.util.*
import kotlin.random.Random

fun generateMap(x: Int, y: Int, platesAmount: Int, random: Random): WorldMap {
    val tiles = createTiles(x, y)
    val map = WorldMap(tiles)
    setTileNeighbours(map)
    val tectonicPlates = randomPlates(
            platesAmount,
            map,
            random
    )
    tectonicPlates.forEach { map.addPlate(it) }
    fill(map)
    return map
}


fun fillResources(
        map: WorldMap,
        resourcePool: ResourcePool,
        supplement: MapGeneratorSupplement,
        random: Random
) {
    for (resource in resourcePool.getResourcesWithPredicate {
        it.spreadProbability != 0.0
                || it.genome.type == Genome.Type.Mineral
    }) {
        scatter(
                map,
                resourcePool,
                resource,
                random.nextInt(
                        supplement.startResourceAmountRange.first,
                        supplement.startResourceAmountRange.last
                ),
                random
        )
    }
}

private fun setTileNeighbours(map: WorldMap) {
    val x = map.x
    val y = map.y
    for (i in 0 until x) {
        for (j in 0 until y) {
            map[i, j].neighbours = arrayOf(
                    map[i, j + 1],
                    map[i, j - 1],
                    map[i + 1, j],
                    map[i - 1, j]
            ).filter { Objects.nonNull(it) }
        }
    }
}

private fun createTiles(x: Int, y: Int): List<MutableList<Tile>> {
    val map: MutableList<MutableList<Tile>> = ArrayList()
    (0 until x).map { i ->
        map.add(
                (0 until y).map { j -> Tile(i, j) }.toMutableList()
        )
    }
    return map
}

private fun randomPlates(platesAmount: Int, map: WorldMap, random: Random): List<TectonicPlate> {
    val tectonicPlates: MutableList<TectonicPlate> = ArrayList()
    val usedTiles: MutableSet<Tile> = HashSet()
    for (i in 0 until platesAmount) {
        val direction = randomElement(
                TectonicPlate.Direction.values().toList(),
                random
        )
        val type = randomElement(
                TectonicPlate.Type.values().toList(),
                random
        )
        val tectonicPlate = TectonicPlate(direction, type)
        val tile = randomTile(map, random)
        tectonicPlate.add(tile)
        tectonicPlates.add(tectonicPlate)
        usedTiles.add(tile)
    }
    var sw = true
    while (sw) {
        sw = false
        for (territory in tectonicPlates) {
            val brink = territory.getBrinkWithCondition { !usedTiles.contains(it) }
            if (brink.isEmpty()) {
                continue
            }
            val tile = randomElement(brink, random)
            territory.add(tile)
            usedTiles.add(tile)
            sw = true
        }
    }
    return tectonicPlates
}

private fun fill(map: WorldMap) {
    var sw = true
    var ssw = true
    for (plate in map.tectonicPlates) {
        if (sw) {
            plate.setType(TectonicPlate.Type.Terrain)
            sw = false
        } else if (ssw) {
            plate.setType(TectonicPlate.Type.Oceanic)
            ssw = false
        }
        plate.initialize()
    }
    map.platesUpdate()
}

private fun scatter(map: WorldMap, resourcePool: ResourcePool, resource: Resource, n: Int, random: Random) {
    val goodTiles = map.getTilesWithPredicate { resource.genome.isOptimal(it) }
    for (i in 0 until n) {
        var tile: Tile
        if (goodTiles.isEmpty()) {
            tile = randomTile(map, random)
            while (!resource.genome.isAcceptable(tile)) {
                tile = randomTile(map, random)
            }
        } else {
            tile = randomElement(goodTiles, random)
        }
        tile.addDelayedResource(resource.copy())
        addDependencies(resource, tile, resourcePool)
    }
}

private fun addDependencies(resource: Resource, tile: Tile, resourcePool: ResourcePool) {
    for (dependency in resource.genome.dependencies) {
        if (!dependency.isPositive || !dependency.isResourceNeeded) {
            continue
        }
        if (dependency is ResourceNeedDependency && dependency.resourceNames.any { it == "Vapour" }) {
            return
        }
        if (dependency is ResourceNeedDependency) {
            for (name in dependency.resourceNames) {
                val dep = resourcePool.getResource(name)
                if (dep.genome.isAcceptable(tile)) {
                    tile.addDelayedResource(dep)
                }
                addDependencies(dep, tile, resourcePool)
            }
            for (name in dependency.materialNames) {
                for (dep in resourcePool.getResourcesWithPredicate {
                    it.spreadProbability > 0
                            && it.simpleName != resource.simpleName
                            && it.genome.primaryMaterial != null
                            && it.genome.primaryMaterial.name == name
                }) {
                    if (dep.genome.isAcceptable(tile)) {
                        tile.addDelayedResource(dep.copy())
                        addDependencies(dep, tile, resourcePool)
                    }
                }
            }
        }
    }
}
